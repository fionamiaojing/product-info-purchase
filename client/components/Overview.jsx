import React from 'react';

export default class Overview extends React.Component {
    constructor(props) {
        super(props);
    }

    render() {
        let info = this.props.info;
        return (
            <ul>
                {info.overview.handmade ? 
                    <li>Handmade item</li> : ""}
                <li>
                    Material: {info.overview.material}
                </li>
                {info.overview['made_to_order'] ? 
                    <li>Made to order</li> : ""}
                <li>
                    Feedback:  
                    <a href="">
                        {info['reviews']} reviews
                    </a>
                </li>
                <li>
                    Favorited by: 
                    <a href="">
                        {info['favorite']} people
                    </a>
                </li>
                {info.overview['gift_message'] ? 
                    <li>Gift message available</li> : ""}
            </ul>
        );
    }
}
