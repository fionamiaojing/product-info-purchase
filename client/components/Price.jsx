import React from 'react';
import _ from 'lodash';

export default class Price extends React.Component {
    constructor(props) {
        super(props);
    }

    render() {
        if (_.isEqual(this.props.originalPrice, this.props.discountedPrice)) {
            return (
                <div id="price">
                <button id="ask">Ask a question</button>
                <span>
                    <span>{this.props.discountedPrice[0]} </span>
                </span>
            </div>
            );
        } else if (this.props.discountedPrice.length === 1 &&
                this.props.originalPrice.length === 1) {
            let discount = 100 - Math.round(
                                this.props.discountedPrice[0] /
                                this.props.originalPrice[0] * 100
                            );
            return (
                <div id="price">
                    <button id="ask">Ask a question</button>
                    <span>
                        <span>{this.props.discountedPrice[0]} </span>
                        <strike>{this.props.originalPrice[0]}</strike>
                        <p>You saved {discount} %</p>
                    </span>
                </div>
            );
        } else {
            return (
                <div id="price">
                    <button id="ask">Ask a question</button>
                    <span>
                        <span>{this.props.discountedPrice[0]}+ </span>
                    </span>
                </div>
            );
        }
        
    }
}