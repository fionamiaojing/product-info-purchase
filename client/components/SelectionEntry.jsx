import React from 'react';

export default class SelectionEntry extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            key: Object.keys(this.props.option)[0],
            value: 'none',

        };
        this.handleChange = this.handleChange.bind(this);
    }

    handleChange(e) {
        this.setState({
            value: e.target.value
        });

        this.props.handleSelect(this.state.key, e.target.value);
    }


    render() {
        return (
            <div>
                <label>
                    {this.state.key}
                </label>
                <span>
                    <div>
                        <select value={this.state.value} onChange={this.handleChange}>
                            {this.state.key === "quantity" ? 
                                '' : <option value="none" >Select an option</option>}
                            
                            {this.props.option[this.state.key].map((value) => 
                                <option key={value} value={value}>{value}</option>
                            )}
                        </select>
                    </div>
                </span>
                <div id="errorSelection" style={{display: 'none'}}>Please select an option</div>
            </div>
        );
    }
}