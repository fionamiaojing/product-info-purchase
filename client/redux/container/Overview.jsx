import React from 'react';
import { connect } from 'react-redux';

class Overview extends React.Component {

    render () {
        return (
            <div>
                <h2>Overview</h2>
                <ul>
                    {this.props.overview.handmade ? 
                        <li>Handmade item</li> : ""}
                    <li>
                        Material: {this.props.overview.material}
                    </li>
                    {this.props.overview['made_to_order'] ? 
                        <li>Made to order</li> : ""}
                    <li>
                        {'Feedback: '} 
                        <a href="#reviews">
                            {this.props.reviews} reviews
                        </a>
                    </li>
                    <li>
                        {'Favorited by: '}
                        <a href="">
                            {this.props.favorite} people
                        </a>
                    </li>
                    {this.props.overview['gift_message'] ? 
                        <li>Gift message available</li> : ""}
                </ul>
            </div>
        );
    }
    
}

function mapStateToProps(state) {
    return {
        overview: state.group.overview,
        reviews: state.group.reviews,
        favorite: state.group.favorite
    };
}

export default connect(
    mapStateToProps
)(Overview);