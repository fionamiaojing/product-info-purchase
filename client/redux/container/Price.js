import React from 'react';
import {bindActionCreators} from 'redux';
import {connect} from 'react-redux';

class Price extends React.Component {
    renderPriceTag() {

    }
}

function mapStateToProps(state) {
    return {
        
    };
}