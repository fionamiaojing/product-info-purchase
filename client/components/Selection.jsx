import React from 'react';
import SelectionEntry from './SelectionEntry.jsx';

export default class Selection extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            //creat an array of 1 - 15
            quantityArray: { quantity: 
                        Array.apply(null, {length: 16}).map(Number.call, Number).slice(1)
            },
            //options for clothes and shoes
            options: {},
            quantity: 1
        };
        this.handleSelect = this.handleSelect.bind(this);
    }

    //mapping options based on category
    mappingOptions(category) {
        if (category === "Shoes" || category === "Clothes") {
            //Shoes and Clothes both have property of color and size
            let properties = ['color', 'size'];
            let options = [];
            //generate options array for color and size respectivity
            //then use the options array to render SelectionEntry
            properties.forEach((property) => {
                let option = {};
                option[property] = this.fetchProperty(property);
                options.push(option);
            });
            return options;
        } else {
            //return empty array
            return [];
        }
    }

    //fetch options based on property
    fetchProperty(property) {
        let output = {};
        this.props.items.forEach((ele) => {
            output[ele[property]] = true;
        });
        return Object.keys(output).sort((a, b) => (
            a - b
        ));
    }

    handleSelect(key, value) {
        if (key === "quantity") {
            this.setState({
                quantity: value
            });
            this.props.detectSelection(value, null);
        } else if (value === "none") {
            // if key already in options
            // delete key from the options;
            if (Object.keys(this.state.options).includes(key)) {
                let newOptions = this.state.options;
                delete newOptions[key];
                this.setState({
                    options: newOptions
                });
                this.props.detectSelection(null, newOptions);
            }
        } else {
            // update options
            let newOptions = this.state.options;
            newOptions[key] = value;
            //set State
            this.setState({
                options: newOptions
            });
            this.props.detectSelection(null, newOptions);
        }
    }

    render() {
        const options = this.mappingOptions(this.props.category);
        return (
            <div id="selections">
                {options.map((option) => 
                    <SelectionEntry 
                        key={Object.keys(option)[0]} 
                        option={option}
                        displayError={this.props.displayError}
                        handleSelect={this.handleSelect}/>
                )}
                {<SelectionEntry 
                    key={'quantity'} 
                    option={this.state.quantityArray}
                    handleSelect={this.handleSelect}/>}
            </div>
        );
    }
}