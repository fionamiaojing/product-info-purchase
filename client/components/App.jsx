import React from 'react';
import axios from 'axios';
import Price from './Price.jsx';
import Selection from './Selection.jsx';
import AlmostGone from './AlmostGone.jsx';
import PeopleWant from './PeopleWant.jsx';
import Overview from './Overview.jsx';
import Shipping from './Shipping.jsx';

export default class App extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            pid: '',
            group: {},
            items: {},
            username: 'hrsf950001',
            options: {},
            quantity: 1,
            selectedItemId: "", //initialize when componentDidMount
            originalPrice: '',
            discountedPrice: '',
            displayError: false,
            shippingInfo: {},
        };
       this.detectSelection = this.detectSelection.bind(this);
       this.handleClick = this.handleClick.bind(this);

    }

    componentDidMount() {
        this.fetchProduct((data) => {
            this.setState({
                group: data.group,
                items: data.items
            });
            if (
                data.group.category === "Handbags" || 
                data.group.category === "Home") {
               let item = data.items[0];
               this.setState({
                   selectedItemId: item['_id'],
                   originalPrice: item['original_price'],
                   discountedPrice: item['discounted_price']
               });
            } else {
               this.setPrices(data.items);
            }
        });
        //fetch shipping information
        this.fetchShippingInfo((data) => {
            this.setState({
                shippingInfo: data
            });
        });
    }

    detectSelection(quantity, options) {
        // update option and quantity
        this.setState({
            options: options || this.state.options,
            quantity: quantity || this.state.quantity
        });
        // assess if all options are provided
        if (!options) {
            //if options not provided, do nothing
            return;
        }
        if (Object.keys(options).length === 2) {
            // set selectedItemID
            let matchItem = this.state.items.filter((item) => {
                return item.color === options.color && item.size === options.size;
            });
            this.setPrices(matchItem);
            let id = matchItem[0]['_id'];
            this.setState({
                selectedItemId: id
            });
        } else if (Object.keys(options).length === 1) {
            let key = Object.keys(options)[0];
            let matchItem = this.state.items.filter((item) => {
                return item[key] === options[key];
            });
            this.setPrices(matchItem);
            //set selectItem to none
            this.setState({
                selectedItemId: ''
            });
        }
    }

    setPrices(items) {
        let originalPrice = {};
        let discountedPrice = {};
        for (var item of items) {
            originalPrice[item['original_price']] = true;
            discountedPrice[item['discounted_price']] = true;
        }
        this.setState({
            originalPrice: Object.keys(originalPrice).map(
                (e) => Number(e)).sort((a, b) => {
                                return a - b;
                }),
            discountedPrice: Object.keys(discountedPrice).map(
                (e) => Number(e)).sort((a, b) => {
                    return a - b;
                }),
        });
    }

    handleClick() {
        if (this.state.selectedItemId === '') {
            this.setState({
                displayError: true
            });
        } else {
            this.send({
                itemId: this.state.selectedItemId,
                quantity:  this.state.quantity 
            });
        }
    }

    send(cartItem) {
        axios.post(`/listing/${this.props.pid}/cart/${this.state.username}`, cartItem)
        //   .then((response) => {
        //       console.log(response.data);
        //   })
          .catch((error) => {
              throw error;
          });
    }

    fetchProduct(callback) {
        axios.get(`/listing/item/${this.props.pid}`)
          .then((response) => {
              callback(response.data);
          })
          .catch((error) => {
              throw error;
          });
    }

    fetchShippingInfo(callback) {
        axios.get(`/listing/${this.props.pid}/shippingInfo`)
          .then((response) => {
              callback(response.data);
          })
          .catch((error) => {
              throw error;
          });
    }

    render() {
        return (
            <div className="sidebar">
                <div id="tophalf">
                    <h1>{this.state.group.title}</h1>
                    <div>
                        <Price 
                            originalPrice={this.state.originalPrice}
                            discountedPrice={this.state.discountedPrice}/>
                        <Selection
                            category={this.state.group.category} 
                            items={this.state.items}
                            displayError={this.state.displayError}
                            detectSelection={this.detectSelection}/>
                    </div>
                </div>
                <button className='buttons' id="buy" onClick={this.handleClick}>Buy it now ></button>
                <button className='buttons' id="add" onClick={this.handleClick}>Add to cart</button>
                <div id="bottom">
                    {this.state.group['number_in_storage'] < 10 ?
                       <AlmostGone amount={this.state.group['number_in_storage']}/> :
                       <PeopleWant amount={this.state.group['count_in_cart']}/>
                    }
                </div>
                <hr/>
                <div id="overview">
                    <h2>Overview</h2>
                    {Object.keys(this.state.group).length > 0 ? 
                        <Overview info={this.state.group}/> :
                        ''
                    }
                </div>
                <hr/>
                <div>
                   <h2>Shipping & returns</h2>
                   <Shipping 
                        shippingInfo={this.state.shippingInfo}
                        pid={this.props.pid}/>
                </div>
                <hr/>
            </div>
        );
    }
}