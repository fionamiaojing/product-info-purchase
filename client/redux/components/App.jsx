import React from 'react';
import axios from 'axios';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { fetchProductData, fetchShippingInfo } from '../action/index';
import Selection from '../container/Selection.jsx';
import Overview from '../container/Overview.jsx';
import Price from '../container/Price.jsx';
import AddButton from '../container/AddButton.jsx';
import DisplayHot from '../container/DisplayHot.jsx';
import Shipping from '../container/Shipping.jsx';
import styles from '../../style.css';

class App extends React.Component {

    componentDidMount() {
        this.props.fetchProductData(pid);
        this.props.fetchShippingInfo(pid);
    }

    render() {
        if (Object.keys(this.props.group).length 
            && this.props.allItems.length
        ) {
            return (
                <div className={styles.sidebar}>
                    <h1 className={styles.h1}>{this.props.group.title}</h1>
                    <div>
                        <Price />
                        <Selection />
                        <AddButton />
                        <DisplayHot />
                        <hr/>
                        <Overview />
                        <hr/>
                        <Shipping />
                    </div>
                </div>
            );
        }
        return (
            <div className={styles.loader}>
                
            </div>
        );
        
    }
}

const pid = document.URL.match(/\/[\w]+\/([\d]+)/)[1];

const matchStateToProps = (state) => {
    return {
        group: state.group,
        allItems: state.allItems
    };
};

const matchDispatchToProps = (dispatch) => {
    return bindActionCreators({ 
        fetchProductData: fetchProductData,
        fetchShippingInfo: fetchShippingInfo,
    }, dispatch);
};

export default connect(
    matchStateToProps,
    matchDispatchToProps
)(App);


// fetchProduct() {
//     axios.get(`/purchase/item/${pid}`)
//       .then((response) => {
//           this.props.receiveProductData(response.data);
//       })
//       .catch((error) => {
//           throw error;
//       });
// }


// fetchShippingInfo() {
//     axios.get(`/purchase/shippingInfo/${pid}`)
//       .then((response) => {
//           this.props.receiveShippingInfo(response.data);
//       })
//       .catch((error) => {
//           throw error;
//       });
// }