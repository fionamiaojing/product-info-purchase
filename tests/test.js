import React from 'react';
import Enzyme, { shallow, render, mount } from 'enzyme';
import { configure } from 'enzyme';
import jest from 'jest-mock';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import App from '../client/components/App.jsx';
import Price from '../client/components/Price.jsx';
import Selection from '../client/components/Selection.jsx';
import SelectionEntry from '../client/components/SelectionEntry.jsx';
import Overview from '../client/components/Overview.jsx';

import data from './testData.js';

configure({ adapter: new Adapter() });

describe('<App />', () => {

    test('it shoud accept a pass-in props', () => {
        const pid = 1001;
        const wrapper = mount(<App pid={pid}/>);
        expect(wrapper.props().pid).toEqual(1001);
    });

    test('should rendering the component', () => {
        const wrapper = mount(<App />);
        expect(wrapper.find('h1')).toHaveLength(1);
    });

    const wrapper = shallow(<App />);

    wrapper.instance().setState({
        group:  data.group
    });

    wrapper.instance().setState({
        items:  data.items
    });

    test('should rendering correct data', () => {
      expect(wrapper.instance().state.group.fake_id).toEqual(1001);
    });

    test('it should detect user selection', () => {
      const quantity = 2;
      const options = {
        "color": "Flesh",
        "size": "5.5",
      };
      wrapper.instance().detectSelection(quantity, options);
      expect(wrapper.instance().state.quantity).toEqual(2);
      expect(wrapper.instance().state.selectedItemId).toEqual(
        '5b135ecbdaf6651723ee8b1c');
    });

});

describe('<Price />', () => {
    let originalPrice = {
        41: true,
    };

    let discountedPrice = {
        36: true,
    };

    const wrapper = mount(
        <Price 
            originalPrice={originalPrice}
            discountedPrice={discountedPrice}
        />
    );
    test('should rendering the component', () => {
        expect(wrapper.find('#price')).toHaveLength(1);
    });

    test('it should display the correct priceTag', () => {
        const component = renderer.create(
            <Price 
            originalPrice={originalPrice}
            discountedPrice={discountedPrice}
            />
        );

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();

    });
});

describe('<SelectionEntry />', () => {
    let option = {
        color: ['red', 'black', 'white']
    };
    
    test('it should display the correct options of colors', () => {
        const component = renderer.create(
            <SelectionEntry 
            option={option}/>
        );
    
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('it should call handleSelect when option is selected', () => {
        const handleSelect = jest.fn();
        const wrapper = mount(
            <SelectionEntry 
            option={option}
            handleSelect={handleSelect}/>
        );
        const button = wrapper.find('#optionSelect');
        button.simulate('change');
        expect(handleSelect).toHaveBeenCalledTimes(1);
    });
});

describe('<Selection />', () => {
    const items = data.items;
    const category = 'Shoes';

    const wrapper = mount(
        <Selection
            items={items}
            category={category}
        />
    );

    test('it should render the correct number of options', () => {
        const numberOfOptions = wrapper.find('#selections').children();
        expect(numberOfOptions.length).toEqual(3);
    });

    test('it should display the correct options for all properties', () => {
        const component = renderer.create(
            <Selection
                items={items}
                category={category}
            />
        );
    
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe('<Overview />', () => {
    let info = {
        overview: {
            "material": "Some kind of material",
            "gift_message": true,
            "made_to_order": true,
            "handmade": true
        },
        reviews: 900,
        favorite: 300
    };

    test('it should display the correct information', () => {
        const component = renderer.create(
            <Overview
                info={info}
            />
        );
    
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});