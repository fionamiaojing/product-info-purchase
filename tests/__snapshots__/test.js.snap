// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Overview /> it should display the correct information 1`] = `
<ul>
  <li>
    Handmade item
  </li>
  <li>
    Material: 
    Some kind of material
  </li>
  <li>
    Made to order
  </li>
  <li>
    Feedback: 
    <a
      href="#reviews"
    >
      900
       reviews
    </a>
  </li>
  <li>
    Favorited by: 
    <a
      href=""
    >
      300
       people
    </a>
  </li>
  <li>
    Gift message available
  </li>
</ul>
`;

exports[`<Price /> it should display the correct priceTag 1`] = `
<div
  id="price"
>
  <div>
    <button
      id="ask"
      onClick={[Function]}
    >
      Ask a question
    </button>
  </div>
  <span
    id="discountPrice"
  >
    $
    + 
  </span>
</div>
`;

exports[`<Selection /> it should display the correct options for all properties 1`] = `
<div
  id="selections"
>
  <div
    id="selection"
  >
    <label>
      color
    </label>
    <span>
      <div>
        <select
          id="optionSelect"
          onChange={[Function]}
          value="none"
        >
          <option
            value="none"
          >
            Select an option
          </option>
          <option
            value="Flesh"
          >
            Flesh
          </option>
          <option
            value="Blue"
          >
            Blue
          </option>
          <option
            value="Beige"
          >
            Beige
          </option>
          <option
            value="Black"
          >
            Black
          </option>
          <option
            value="Champagne"
          >
            Champagne
          </option>
          <option
            value="White"
          >
            White
          </option>
          <option
            value="Gray"
          >
            Gray
          </option>
          <option
            value="Red"
          >
            Red
          </option>
        </select>
      </div>
    </span>
    <div
      id="errorSelection"
      style={
        Object {
          "display": "none",
        }
      }
    >
      Please select a 
      color
    </div>
  </div>
  <div
    id="selection"
  >
    <label>
      size
    </label>
    <span>
      <div>
        <select
          id="optionSelect"
          onChange={[Function]}
          value="none"
        >
          <option
            value="none"
          >
            Select an option
          </option>
          <option
            value="5.5"
          >
            5.5
          </option>
          <option
            value="6"
          >
            6
          </option>
          <option
            value="6.5"
          >
            6.5
          </option>
          <option
            value="7"
          >
            7
          </option>
          <option
            value="7.5"
          >
            7.5
          </option>
          <option
            value="8"
          >
            8
          </option>
          <option
            value="8.5"
          >
            8.5
          </option>
          <option
            value="9"
          >
            9
          </option>
          <option
            value="9.5"
          >
            9.5
          </option>
          <option
            value="10"
          >
            10
          </option>
          <option
            value="10.5"
          >
            10.5
          </option>
        </select>
      </div>
    </span>
    <div
      id="errorSelection"
      style={
        Object {
          "display": "none",
        }
      }
    >
      Please select a 
      size
    </div>
  </div>
  <div
    id="selection"
  >
    <label>
      quantity
    </label>
    <span>
      <div>
        <select
          id="optionSelect"
          onChange={[Function]}
          value="none"
        >
          
          <option
            value={1}
          >
            1
          </option>
          <option
            value={2}
          >
            2
          </option>
          <option
            value={3}
          >
            3
          </option>
          <option
            value={4}
          >
            4
          </option>
          <option
            value={5}
          >
            5
          </option>
          <option
            value={6}
          >
            6
          </option>
          <option
            value={7}
          >
            7
          </option>
          <option
            value={8}
          >
            8
          </option>
          <option
            value={9}
          >
            9
          </option>
          <option
            value={10}
          >
            10
          </option>
          <option
            value={11}
          >
            11
          </option>
          <option
            value={12}
          >
            12
          </option>
          <option
            value={13}
          >
            13
          </option>
          <option
            value={14}
          >
            14
          </option>
          <option
            value={15}
          >
            15
          </option>
        </select>
      </div>
    </span>
    <div
      id="errorSelection"
      style={
        Object {
          "display": "none",
        }
      }
    >
      Please select a 
      quantity
    </div>
  </div>
</div>
`;

exports[`<SelectionEntry /> it should display the correct options of colors 1`] = `
<div
  id="selection"
>
  <label>
    color
  </label>
  <span>
    <div>
      <select
        id="optionSelect"
        onChange={[Function]}
        value="none"
      >
        <option
          value="none"
        >
          Select an option
        </option>
        <option
          value="red"
        >
          red
        </option>
        <option
          value="black"
        >
          black
        </option>
        <option
          value="white"
        >
          white
        </option>
      </select>
    </div>
  </span>
  <div
    id="errorSelection"
    style={
      Object {
        "display": "none",
      }
    }
  >
    Please select a 
    color
  </div>
</div>
`;
